### ISSUE TITLE
Proposal: tutorial for sample-stable image augmentation across workers and resume

### ISSUE THREAD

**Issue by ternaus** [huggingface/accelerate#4191]
Hello Accelerate maintainers,

Would you be open to an optional tutorial showing how to make each image augmentation depend on the sample and epoch, rather than on whichever worker happens to request the sample first?

The current [`complete_cv_example.py`](https://github.com/huggingface/accelerate/blob/main/examples/complete_cv_example.py) is a good setting for this problem: it trains on Oxford-IIIT Pets with four DataLoader workers, distributed preparation, checkpointing, and resume support. Accelerate can make the sampler sequence reproducible and can restore DataLoader position with `use_stateful_dataloader`. A third-party augmentation pipeline, however, can still own a separate random stream. If its next crop or flip is selected by call order, changing the worker/process schedule or resuming through a different path can give the same sample a different augmentation.

This is not an Accelerate bug, and the proposed tutorial would keep Accelerate's current seeding, sampler, and DataLoader-state mechanisms. It would document the remaining boundary: Accelerate decides which sample comes next, while the augmentation pipeline decides which transformation that sample receives. The existing [internal-mechanism guide](https://github.com/huggingface/accelerate/blob/main/docs/source/concept_guides/internal_mechanism.md) already notes that RNG synchronization can affect random data augmentation; the tutorial would turn that warning into a runnable image example.

## Proposed tutorial

I suggest a focused variant of the existing Pets example:

1. Keep the model, optimizer, `Accelerator.prepare()`, checkpointing, resume flow, and deterministic validation transform unchanged.
2. Use the image filename as a stable sample ID and combine it with the base seed and epoch.
3. Pass the derived value as AlbumentationsX's per-call `invocation_seed`, so the same `(epoch, sample_id)` receives the same crop, flip, and color realization regardless of call order.
4. Finish the AX pipeline with `ToFloat` and `ToTensorV2`.
5. Record a compact manifest of `(epoch, sample_id, augmented_image_digest)` and compare an uninterrupted run with a checkpoint/resume run, `num_workers=0` with `num_workers=4`, and a single-process run with a two-process launch.

The core training-only Dataset variant would look like this; the existing `PetsDataset` and deterministic torchvision transform can remain in use for evaluation:

```python
import hashlib
import os

import albumentations as A
import numpy as np
from albumentations.pytorch import ToTensorV2


def augmentation_seed(base_seed: int, epoch: int, sample_id: str) -> int:
    payload = f"{base_seed}:{epoch}:{sample_id}".encode()
    return int.from_bytes(hashlib.blake2b(payload, digest_size=8).digest(), "little")


base_seed = seed
train_tfm = A.Compose(
    [
        A.RandomResizedCrop(size=image_size, scale=(0.5, 1.0), p=1.0),
        A.HorizontalFlip(p=0.5),
        A.RandomBrightnessContrast(p=0.5),
        A.ToFloat(max_value=255.0),
        ToTensorV2(),
    ],
    strict=True,
    seed=base_seed,
)


class SampleStablePetsDataset(Dataset):
    def __init__(self, file_names, image_transform=None, label_to_id=None, base_seed=0):
        self.file_names = file_names
        self.image_transform = image_transform
        self.label_to_id = label_to_id
        self.base_seed = base_seed
        self.epoch = 0

    def __len__(self):
        return len(self.file_names)

    def set_epoch(self, epoch: int) -> None:
        self.epoch = epoch

    def __getitem__(self, idx):
        fname = self.file_names[idx]
        sample_id = os.path.basename(fname)
        image = PIL.Image.open(fname).convert("RGB")

        if self.image_transform is not None:
            result = self.image_transform(
                image=np.asarray(image),
                invocation_seed=augmentation_seed(self.base_seed, self.epoch, sample_id),
            )
            image = result["image"]

        label = extract_label(fname)
        if self.label_to_id is not None:
            label = self.label_to_id[label]
        return {"image": image, "label": label}


train_dataset = SampleStablePetsDataset(
    [file_names[i] for i in train_split],
    image_transform=train_tfm,
    label_to_id=label_to_id,
    base_seed=base_seed,
)
```

The training loop would call `train_dataset.set_epoch(epoch)` before creating that epoch's DataLoader iterator. This matches the current example's non-persistent workers. The tutorial would call out that persistent workers need an epoch value shared with worker processes rather than a normal Dataset attribute.

## Expected result

For a fixed base seed, the digest for a given `(epoch, sample_id)` should match across worker counts, process assignment, and checkpoint/resume. A different epoch should produce a new but repeatable realization. This isolates the input augmentation path; it would not claim bitwise-identical model training across different hardware or kernels.

I locally checked the core property with AlbumentationsX 2.4.3: reversing sample call order preserved the exact output tensor for each derived per-sample seed, while using only a stateful pipeline seed changed every sample when the call order was reversed.

## Scope and placement

This would be an optional documentation/example dependency, not an Accelerate core API change or default dependency. AlbumentationsX is installed as `albumentationsx` and imported as `albumentations`; it is AGPL-3.0-only, requires Python 3.10+, and expects users to choose the appropriate PyTorch installation for CPU, CUDA, or MPS.

If the topic is useful, I would be happy to prepare it in whichever form is easiest to maintain: a focused section around the complete CV example, an `examples/by_feature` example, or a short standalone reproducibility guide. I would wait for your preferred scope and placement before opening a pull request.

**Comment by harshitakrishna18** [comment]
@ternaus Personally I think this could be a great addition. Additionally if you want to tag team and work on creating these documentations/examples, I would be happy to help!

**Comment by SunMarc** [comment]
Thanks for the proposition @ternaus ! I prefer to not introduce optional documentation dependency in accelerate. I encourage you instead to write a HF blog as the reach will probably be better and more suited for this kind of thing ! https://huggingface.co/blog
